src/Arcus.BackgroundJobs.AzureActiveDirectory/ClientSecretExpirationJobOptions.cs

```diff
@@ -0,0 +1,109 @@
+        /// </summary>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown when the <paramref name="value"/> is less than zero or greater than 23.</exception>
+        public int RunAtHour
+        {
+            get => _runAtHour;
+            set
+            {
+                Guard.NotLessThan(value, 0, nameof(value));
+                Guard.NotGreaterThan(value, 23, nameof(value));
+
+                _runAtHour = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets whether the task should run immediately, this should only be used for testing purposes.
+        /// </summary>
+        public bool RunImmediately
+        {
+            get => _runImmediately;
+            set
+            {
+                _runImmediately = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the uri of the CloudEvent that will be published to Event Grid.
+        /// </summary>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown when the <paramref name="value"/> is blank.</exception>
+        public Uri EventUri
+        {
+            get => _eventUri;
+            set
+            {
+                Guard.NotNullOrWhitespace(value.OriginalString, nameof(value));
+
+                _eventUri = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the threshold for the expiration, if the end datetime for a secret is lower than this value a <see cref="CloudEvent"/> will be published.
+        /// </summary>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown when the <paramref name="value"/> is less than zero.</exception>
+        public int ExpirationThreshold
+        {
+            get => _expirationThreshold;
+            set
+            {
+                Guard.NotLessThan(value, 0, nameof(value));
+
+                _expirationThreshold = value;
+            }
+        }
+
+        /// <summary>
+        /// Creates a <see cref="CloudEvent"/> instance using the predefined values.
+        /// </summary>
+        /// <param name="application">The <see cref="AzureApplication"/> containing the information regarding the application and its expiring or about to expire secret.</param>
+        /// <param name="type">The type used in the creation of the <see cref="CloudEvent"/>.</param>
+        /// <param name="eventUri">The uri used in the creation of the <see cref="CloudEvent"/>.</param>
+        /// <exception cref="ArgumentNullException">Thrown when the <paramref name="application"/> is null.</exception>
+        /// <exception cref="ArgumentException">Thrown when the <paramref name="application"/> name is blank.</exception>
+        /// <exception cref="ArgumentException">Thrown when the <paramref name="eventUri"/> is blank.</exception>
+        public virtual CloudEvent CreateEvent(AzureApplication application, ClientSecretExpirationEventType type, Uri eventUri)
```

Comment:
We might consider adding an method overload `AddClientSecretExpirationJob<TOptions>` so consumers can create their own implementation of the options and override this method. Otherwise, it's not really useful.

On the other hand, we may consider not making this virtual and keeping it internal for now, until their is an demand for it.

Reply:
I think both options are fine, but not sure how much demand their would be to override this method, maybe keep it internal for now? @stijnmoreels What do you think is the best approach?

Reply:
Yeah, I was thinking to keep it internal for now. We can always come back on it later.

Reply:
Changed it, see [`db6e6a6` (#87)](https://github.com/arcus-azure/arcus.backgroundjobs/pull/87/commits/db6e6a61135c442a7d301cfb600d2d48ab310f46)